### ISSUE TITLE
ConcurrentBattleCalculator#calculate:200 - java.lang.IllegalStateException

### ISSUE THREAD

**Issue by tripleabuilderbot** [triplea-game/triplea#12488]
## Map
world_war_ii_global

## Log Message
java.lang.IllegalStateException: Round 10,000 reached in a battle. Something must be wrong. Please report this to TripleA.
 Territory: 62 Sea Zone Attacker: Japanese Attacking unit types: carrier,battleship,fighter,transport,destroyer,cruiser, Defending unit types: carrier,battleship,submarine,fighter,destroyer,cruiser

## TripleA Version
[2.6.14470](https://github.com/triplea-game/triplea/releases/tag/2.6.14470)

## Java Version
11.0.4

## Operating System
Windows 10

## Stack Trace
```
Exception: java.lang.IllegalStateException java.lang.IllegalStateException: Round 10,000 reached in a battle. Something must be wrong. Please report this to TripleA.
 Territory: 62 Sea Zone Attacker: Japanese Attacking unit types: carrier,battleship,fighter,transport,destroyer,cruiser, Defending unit types: carrier,battleship,submarine,fighter,destroyer,cruiser
java.lang.Exception
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(Unknown Source)
	at java.base/java.lang.reflect.Constructor.newInstance(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.getThrowableException(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.reportException(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.invoke(Unknown Source)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateParallel(Unknown Source)
	at java.base/java.util.stream.AbstractPipeline.evaluate(Unknown Source)
	at java.base/java.util.stream.ReferencePipeline.collect(Unknown Source)
	at games.strategy.triplea.odds.calculator.ConcurrentBattleCalculator.calculate(ConcurrentBattleCalculator.java:200)
	at games.strategy.triplea.odds.calculator.BattleCalculatorPanel.lambda$updateStats$13(BattleCalculatorPanel.java:1173)
	at java.base/java.lang.Thread.run(Unknown Source)


Exception: java.lang.IllegalStateException Round 10,000 reached in a battle. Something must be wrong. Please report this to TripleA.
 Territory: 62 Sea Zone Attacker: Japanese Attacking unit types: carrier,battleship,fighter,transport,destroyer,cruiser, Defending unit types: carrier,battleship,submarine,fighter,destroyer,cruiser
java.lang.Exception
	at games.strategy.triplea.delegate.battle.MustFightBattle$29.execute(MustFightBattle.java:1344)
	at games.strategy.triplea.delegate.ExecutionStack.execute(ExecutionStack.java:34)
	at games.strategy.triplea.delegate.battle.MustFightBattle.fight(MustFightBattle.java:714)
	at games.strategy.triplea.odds.calculator.BattleCalculator.calculate(BattleCalculator.java:115)
	at games.strategy.triplea.odds.calculator.ConcurrentBattleCalculator.lambda$calculate$4(ConcurrentBattleCalculator.java:188)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(Unknown Source)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Unknown Source)
	at java.base/java.util.stream.AbstractPipeline.copyInto(Unknown Source)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source)
	at java.base/java.util.stream.ReduceOps$ReduceTask.doLeaf(Unknown Source)
	at java.base/java.util.stream.ReduceOps$ReduceTask.doLeaf(Unknown Source)
	at java.base/java.util.stream.AbstractTask.compute(Unknown Source)
	at java.base/java.util.concurrent.CountedCompleter.exec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)


```

**Comment by beelee1** [comment]
hmm .. havent seen this in 2.6 Maybe try the latest 2.6 It might be already fixed

**Comment by fasthard** [comment]
Thx for your answer, but isn't 2.6.14470 the latest 2.6 version?

**Comment by beelee1** [comment]
@fasthard looks as if 14705 is latest. https://github.com/triplea-game/triplea/releases/

I've been using 14696 and haven't noticed it. It may have just not come up yet though. Looked like when same units went at it is when it nutted up. So that's possible it's not known about and still going on.

**Comment by asvitkine** [comment]
@fasthard do you remember what happened when you got this error? Do you happen to have an autosave from a phase right before this that reproduces the error after loading?

We've gotten a few reports of this type of error, but not a reliable way to reproduce, so any help would be appreciated!

**Comment by fasthard** [comment]
Hi, well I edited this file:
[rus006.zip](https://github.com/triplea-game/triplea/files/14956022/rus006.zip)
to this file:
[BM4_O&EvsA+22_US7_.zip](https://github.com/triplea-game/triplea/files/14956028/BM4_O.EvsA%2B22_US7_.zip)
Thanks for you effort!

**Comment by asvitkine** [comment]
@fasthard 

Those two saves have all players set to human and there are no pending combat moves. So it doesn't seem like there's enough to reproduce the issue there?

Can you clarify again how you got the error? Were you playing this game online with a bunch of humans - or did you play with AIs? You don't have any autosaves from right before the bug happened?

**Comment by asvitkine** [comment]
Although I guess I can reproduce this by just leaving edit mode on when ending a phase.
Here's a repro:

[autosaveBeforeEndTurn.tsvg.zip](https://github.com/triplea-game/triplea/files/15051744/autosaveBeforeEndTurn.tsvg.zip)

I guess we can make it so ending a phase will turn off edit mode. Or we can make it so no combats happen in edit mode.

**Comment by asvitkine** [comment]
Looks like the error is not new to 2.6. I tried with 2.5 and hit this one: https://github.com/triplea-game/triplea/issues/12434

**Comment by asvitkine** [comment]
I looked through at the code history of when edit mode was made to ignore combat casualties and it was when it was originally added here:
https://github.com/triplea-game/triplea/commit/ab56e9cc32ebf0eb8bb2d32c8204a45839e7d113

So I guess it always worked this way (i.e. no casualties are done when edit mode is active in combat). I'm actually not sure why that's the case. To me, edit mode is just about editing the state of the game. I feel like after you do that, there's no reason to also make combat dice rolls have no effect - why not let it happen as normal?

That would simplify the code a lot and let us get rid of bugs like these.

**Comment by beelee1** [comment]
@asvitkine 

yea, so the G 40 Expansion mod uses edit with no casualties being selected so it can produce the desired game result that triplea can't do otherwise.

For example, there are units that get a +1 on 1st rd of attack or defense and then revert to there normal combat modes. The only way I can make it work, is to play the battle in edit and pick the casualties manually. They still show as a hit or not according to there combat values but don't auto hit.

If you make it so they auto hit, it will require a mass amount of edit to use and since the dice roller isn't documented, which would be a slower process, one would have to use an outside dice roller for verification.

Please don't change this behavior or maybe I have misunderstood what was being discussed

**Comment by Cernelius** [comment]
> So I guess it always worked this way (i.e. no casualties are done when edit mode is active in combat). I'm actually not sure why that's the case. To me, edit mode is just about editing the state of the game. I feel like after you do that, there's no reason to also make combat dice rolls have no effect - why not let it happen as normal?

I don't think so. I remember that in the past having Edit Mode selected just allowed you to take whatever casualties you wanted (comprising possibly none at all), then (many years ago) the behaviour changed to everyone being immortal with Edit Mode on.

I guess it was not an intended change, but I think I've never reported it.

It was many years ago when this happened, so I might be remembering wrong, but I'm almost sure of what I'm saying.

**Comment by asvitkine** [comment]
@beelee1 

Interesting. Can you give more details about how selecting casualties manually works in edit mode? From what I can tell (before my change) is it would just default to no units taking damage, but the dialog to select which ones to take as casualties still pops up.

Do you mean that you still play through the combat, but after each roll of dice, you go back to the map and delete units (and that affects the running battle)? Or is there some way to do this from the combat UI?

Seems like a pretty cumbersome use case. Perhaps we can just support whatever rules "G 40 Expansion" required natively in the engine? Are there any maps where people need to do this?

**Comment by beelee1** [comment]
@asvitkine  yea so here's how I currently do it on 2.6 14705 Just a quick hack to show it. If there are 3 Submarines, they get A1 D1 for first rd of combat only. Then they get there normal combat values of A2 D1. Here are some shots

so we have 3 U-boats attacking. When there are 3 they get a +1 attack for first rd only. So i use edit to be able to pick the casualties as needed.

![Screenshot from 2024-04-26 10-56-31](https://github.com/triplea-game/triplea/assets/24382682/f80e8d3c-ee38-477f-aa70-d449d6d8250c)

so no hits in this case as a 4 and two sixes are rolled.

![Screenshot from 2024-04-26 10-59-09](https://github.com/triplea-game/triplea/assets/24382682/e49e3cea-1d82-4c6e-acd4-07e3b44b6e20)

this is what shows up in the battle window. The hits are currently zero but I can pick as many as I want.

![Screenshot from 2024-04-26 11-00-45](https://github.com/triplea-game/triplea/assets/24382682/97d93bf0-0890-43eb-bfff-aaca96babe54)

That way, if there was a hit at 3 in the first rd I could select it. 

Hmm I don't think I explained it very well. Basically you can choose to take a hit or not

**Comment by beelee1** [comment]
@asvitkine yea it's not ideal play wise but only way to be rule compliant with this mod :) It's actually just part of game play at this point as were on game 4 using it.

I'm not entirely sure I'm understanding your change. I haven't tried it yet. On 14705

**Comment by asvitkine** [comment]
@beelee1 

I see, so if you select things in the battle window, then the hits will actually be done?
OK, thanks. My change probably breaks this, so I suppose it should be reverted.

But I can take a look to see if engine support (for new XML) for the above could be added, so edit mode wouldn't be required.

I'm also wondering, I guess the above (edit mode to workaround triplea limitation) only works if the number of dice rolled is the same in the custom rules? For example, if custom rules say "roll an extra die on first round" or something, then it wouldn't work right?

For example, I was looking at that ruleset's pdf:
https://drive.google.com/file/d/1SRDOsOwlO4Aa13cULwkWrorrXEmCmiN-/view

And on page 8, it's talking about Tank Armies and Army Corps which sound like they might be changing the number of dice rolled, etc. Wouldn't that not be supported with how Edit mode combat works due to number of dice differences?

**Comment by beelee1** [comment]
@asvitkine 

yea i'm not sure about the extra die rolls. If they show up in edit or not. I'll test it.

Try not to use edit if at all possible, so basically there are two units, U-boats and Paratroopers that need to be rolled in edit.

Would be way cool if you could add that ability to triplea. Idk how much labor would be involved on your part though.

While not perfect, it currently works as is. 

Also, there may be other games that would need to use edit if they wanted. For example, the 1 DD blocks all subs in the Global game.

But yea, it is cumbersome and mistakes are easier to make, but basically it plays like a tabletop game when using it

**Comment by Cernelius** [comment]
> @asvitkine yea so here's how I currently do it on 2.6 14705 Just a quick hack to show it. If there are 3 Submarines, they get A1 D1 for first rd of combat only. Then they get there normal combat values of A2 D1. Here are some shots
> 
> so we have 3 U-boats attacking. When there are 3 they get a +1 attack for first rd only. So i use edit to be able to pick the casualties as needed.

You mean that they get an ADDITIONAL A1 D1 for the first round of combat only, meaning that they'll have Attack 3 or Defence 2 on the first round and Attack 2 or Defence 1 on whatever other rounds?

@asvitkine A way to do this (which I think would stand a high chance to be used by other games too) would be by allowing an additional option for the current support attachments, defining the number of combat rounds for which the support lasts. Then, the game could be updated by having the submarines giving support to themselves accordingly.

I mean something like this.
```
		<attachment name="supportAttachmentSumbarine" attachTo="submarine" javaClass="games.strategy.triplea.attachments.UnitSupportAttachment" type="unitType">
			<option name="unitType" value="submarine"/>
			<option name="faction" value="allied"/>
			<option name="side" value="offence:defence"/>
			<option name="dice" value="strength"/>
			<option name="bonus" value="1"/>
			<option name="number" value="1"/>
			<option name="bonusType" value="submarine"/>
			<option name="maxRounds" value="1"/>
			<option name="players" value="..."/>
		</attachment>
```

The option which would need to be added is
`<option name="maxRounds" value="1"/>`
.

Descriptively
`maxRounds 						values: Sets for how many rounds the support is given. Negative (-1) means infinite. (default = -1).`
.

I think that such an option may have good applications to several over cases. For example, I can see game-makers wanting to have artillery support lasting only for the first round of combat or somesuch (as that would make sense).

---

While at it, I also suggest adding an "∀" or (if "∀" is not a valid character) "a" (meaning "all") or maybe "+" value for the "players" option, so you don't have to input every player when you want everyone to have the support (like it happens now and sometimes game-makers forget some players, especially the null ("Neutral") pseudo-player).

**Comment by beelee1** [comment]
@Cernelius yea it would be a nice feature that a lot of people would use. Hopefully asvikine is up for it :)

Pretty close, except the Subs get a D+2 so would D3 1st rd and D1 after

**Comment by Cernelius** [comment]
> Pretty close, except the Subs get a D+2 so would D3 1st rd and D1 after

In this case, under my proposal, two attachments would be needed.
```
		<attachment name="supportAttachmentSumbarineOffence" attachTo="submarine" javaClass="games.strategy.triplea.attachments.UnitSupportAttachment" type="unitType">
			<option name="unitType" value="submarine"/>
			<option name="faction" value="allied"/>
			<option name="side" value="offence"/>
			<option name="dice" value="strength"/>
			<option name="bonus" value="1"/>
			<option name="number" value="1"/>
			<option name="bonusType" value="submarine"/>
			<option name="maxRounds" value="1"/>
			<option name="players" value="..."/>
		</attachment>

		<attachment name="supportAttachmentSumbarineDefence" attachTo="submarine" javaClass="games.strategy.triplea.attachments.UnitSupportAttachment" type="unitType">
			<option name="unitType" value="submarine"/>
			<option name="faction" value="allied"/>
			<option name="side" value="defence"/>
			<option name="dice" value="strength"/>
			<option name="bonus" value="2"/>
			<option name="number" value="1"/>
			<option name="bonusType" value="submarine"/>
			<option name="maxRounds" value="1"/>
			<option name="players" value="..."/>
		</attachment>
```

**Comment by Cernelius** [comment]
> @Cernelius yea it would be a nice feature that a lot of people would use. Hopefully asvitkine is up for it :)
> 
> Pretty close, except the Subs get a D+2 so would D3 1st rd and D1 after

So the rule is that submarines are offence/defence 3/3 on the first round of combat and offence/defence 2/1 thereafter?

**Comment by beelee1** [comment]
> So the rule is that submarines are offence/defence 3/3 on the first round of combat and offence/defence 2/1 thereafter?

yes

Yea you'd need one for offense and defense both. The supportAttachment

Would be a super cool feature. There would be a lot of uses for it :)

**Comment by Cernelius** [comment]
I've just read that also air-transported paratroopers have a similar bonus, but I don't know if (in your game) there are dedicated transports for paratroopers (as in reality) or the bombers paratroop and bomb at the same time like in the basic games or paratroopers just jump on their own (not needing any aircraft) like in the 1940 series.

If the paratroopers are transported by air units, you would also need an other option which limits support only to transported units, then you could have the transporting aircraft giving support to the paratroopers (sort of like artillery does to infantry in the basic games). Such an option could be maybe also used for limiting support from land transports only to the transported units and maybe even for having sea transports giving support to the units which they offload. I don't actually know, though.

Definitely sounds good to me that paratroopers being used as such should be able to have different powers than paratroopers used as simple land units (sort of like the Marine bonus), but it's up to @asvitkine or whatever developer to see if they want to support such rules. Guess they'll have to read the rulebook to be sure.

Can you please reference the rulebook and the page or pages where the relevant rules are to be found?

**Comment by beelee1** [comment]
@Cernelius here's the rulebook https://www.axisandallies.org/forums/topic/36945/axis-allies-global-1940-house-rules-expansion

Paratroopers are on page 18 and U-boats page 41. 

Yea the Paratroopers are self supported, so the air transport is included with the unit, similar to G40.

Yea, really hope he takes it on. It'd be way usefull for more than just this mod :)

**Comment by Cernelius** [comment]
I see that paratroopers need to be both alone (beside aa guns) and having moved by air when attacking to have the special offensive or defensive power (attacking at 3 instead of at 1 or defending at 3 instead of at 2 during only the first round of combat). Moreover, paratroopers cannot be supported by artillery if they moved by air.

Assuming supports could be given on the first round of combat only (as per my proposal), the alone part could be actually sort of hacked by giving every fighting land and air non-paratrooper unit beside aa guns a -2 offensive modifier and -1 defensive modifier to infinite allied paratroopers units for the first round of combat while paratroopers give +2 and +1 respectively for offence and defence to themselves for the first round of combat, but this feels a bit of a work-around to me and would not cover the fact that every paratrooper can either move as a movement 1 land unit or as a movement 2 (3 with airfield) air unit which can land in enemy territories. This last alternative movement is sort of already supported in TripleA thanks to the "Airborne Combat Move" (but in the basic 1940.II series is tech dependant and requires the presence of an airfield in the starting territory, whereas no airfield is needed according to these rules but gives total air movement of 3 to paratroopers moving by air if present), comprising the fact that paratroopers moving as air units may get shot by AA guns, but TripleA itself would need to be made able to discern between paratroopers which moved by land and paratroopers which moved by air and somehow confer the offensive bonus only to the latter and only on the first round of combat.

Also considering that it is needed to disable the support ability of artillery in favour of paratroopers if these paratroopers moved by air, I would say this should be done by an additional option to the support attachment too. This option should be a flag which allows/disallows support depending on whether the receiving units moved by air or by other means, and then the artillery support should be updated as not to apply to units which moved by air, whereas two more couples of support attachments would be added, by which every paratrooper gives +2 attack / +1 defence power to itself on the first round of combat, and every non-paratrooper gives -2 attack / -1 defence to infinite paratroopers on the first round of combat, and both of these two additional couples of support attachments do not apply to units which have not moved by air.

Of course, the aforementioned workaround itself could be avoided by having an other option which disallows the support if any other units beside a list of exceptions (like aa guns) is present on its own side (comprising co-defenders, which are usually but not always allies). That way, only the couple of +2/+1 support to self for paratroopers would be needed, with the additional three options that these supports are active only for the first round of combat, only for paratroopers which moved by air and only as long as no other units but paratroopers and aa guns are partaking on the same side.

By the way, it feels very strange to me that paratroopers cannot attack at 3 if they are attacking together with air units like a bomber, but of course rules are rules if I'm understanding that correctly (that, if I send 1 paratrooper alone via air, it attacks at 3, whereas, if I send 1 paratrooper and 1 fighter alone via air, the paratrooper attacks at 1).

Moreover, I guess that the rule
>When a Paratrooper attacks by land, it attacks as
a regular infantry unit at “1” – in this case it can
also be supported by an artillery unit and thereby
attack at “2”.

is incomplete, and it actually applies also in case the paratrooper is offloaded (by a sea transport) so practically in every case the paratrooper attacks not by air (and not just in the case it attacks by land).

Moreover the rule
>If Paratroopers attack by air and are supported by
units other than Paratroopers, the Paratroopers
attack at “1”.

is quite unclear in the moment it states "are supported by units other than Paratroopers", and I assume that what this part of the rules actually means is "are attacking together with units other than Paratroopers", which anyway would not clarify if my aforementioned paratrooper (which is attacking together with 1 fighter and no other units) would attack at 3 if the fighter is shot down by aa guns or anyway eliminated before the paratrooper can make its attack (assuming that the paratrooper attacks at 1 if the fighter attacks too).

Moreover, the aforementioned rule is still likely literally wrong with respect to what I can imagine the author intended to rule because it would imply that, for example, if I send 1 paratrooper by air and 1 paratrooper by land into the same territory to make battle, while sending no other units, then the paratrooper sent by air attacks at 3 (because it is not supported/accompanied by units other than paratroopers), and this clearly makes no sense in the moment that the same paratrooper would instead attack at 1 if the paratrooper moving by land was instead an infantry. My assumption is that the author intended to say that paratroopers attack at 3 only if they are alone and each of them moved by air, but this is not actually stated in these (additional) rules. Of course, I may be wrong here: maybe the author actually wanted that, when you have one or more paratroopers attacking a territory by air together with only one or more other paratroopers attacking the same by land or sea or both, the former paratroopers attack at 3 (on the first round of combat) and the latter ones attack at 1 (which is what the rules are literally stating).

Honestly, I find these paratroopers rules a bit too on the heavy side as a matter of quantity of rules, but that tends to be so with paratroopers rules in general (as they are the most complex tech also in the basic games).

By the way, the rule that paratroopers must be completely alone to have the bonus definitely make no sense to me, but that is just my opinion.

I also dislike the exception by which paratroopers cannot air attack capitals because it is a restriction which makes realistically no sense to me, but that is again just my opinion (and yet an other special rule which would have to be supported which I've previously ignored because I already felt over-loaded).

---

Definitely this is a much more complex and exceptional rule than the simple bonus first round offence/defence for submarines, so **maybe it is better just to think about it only after the submarine rule gets done (if it gets done), so we know at least how that actually works.**

On top of that, if we are close to releasing 2.6, maybe it is better not to add features until after the release for safety? I've no idea about this nor how close to the release we actually are, so I would defer to @asvitkine or any developer.

---

To summarize, my suggestion to @asvitkine or any developer would be to support the aforementioned rules for submarines (units having a simple modifier for the first round of combat only) but to hold off for paratroopers until their rules are made clearer and simpler. However, I would say that having the ability to give paratroopers which are actually being air-dropped a different offensive value than paratroopers which have moved by land (or sea) is something which I see as very valuable (and it is quite surprising to me that the basic games do not have any such rules as far as I know) and conceptually akin to the already supported "Marine" rule.

**Comment by beelee1** [comment]
@Cernelius yea I think the can't attack Capitals is for play balance. You'd be forced to keep a lot of defense there, although a lot of builds are in your Capital anyways. Especially moscow. But that is my understanding of it.

azimuth came up with a workaround here I think it should work. Haven't tried it yet. https://forums.triplea-game.org/topic/3795/broken-air-battles-with-2-6/14

![Screenshot from 2024-04-27 08-59-33](https://github.com/triplea-game/triplea/assets/24382682/0ddb87b7-8182-4046-b5d6-7d856220d842)

**Comment by Cernelius** [comment]
That is not going to work because that way you cannot take the submarine as casualty in the first round of combat (the submarine will become practically invulnerable in the first round of combat).

**Comment by beelee1** [comment]
oh ... bummer :(

yea they'd auto spawn to the new unit hmm ... yea I guess that won't work. Maybe you could have a token that gives the new unit combat support via support attachment and if it's not there the new unit wouldn't be able to do combat ? That way the token could take any first rd hits

no that wouldn't work either besides being a massive hack.

Well hopefully asvitikine takes it on. It's actually not that big of a deal todo it in edit currently. We're kinda used to it at this point :)

**Comment by asvitkine** [comment]
If the current engine could already be scripted via XML to support it, then I don't think new features are needed.

How about having a trigger that spawns a special unit that doesn't have any attack, but provides bonus to 3 submarines on attack and have that unit be flagged as `isSuicideOnAttack`.

Then, you just need to spawn that unit anywhere there are three groups of submarines of the current player. It's a bit convoluted, but I think it would work? The trigger should run at end of combat move phase. And there should be another trigger to remove all of these at the end of combat.

**Comment by beelee1** [comment]
That sounds like it would work. I'll give it a shot

**Comment by beelee1** [comment]
@asvitkine yea unfornutaely isSuicideOnAttack doesn't work. What happens is, that it suicides before the units it's supporting roll. With set to true it didn't work. Set to false, it did work, but of course it doesn't kill itself.

That was good idea though and worth a try.

**Comment by beelee1** [comment]
@asvitkine well it works on defense with one exception. This a G 40 mod and if submarines are attacked by other submarines with a Destroyer and they don't have a Destroyer to protect them, they will lose any FirstStrike hits before they can fire back.

This could lead to a situation where the Wolfpack would disolve due to having less than 3 subs. But they would still get there bonus on first rd of defense. Which would be incorrect.

Not a huge issue, but one would need to use edit in that case.

Still a great improvement :)

**Comment by beelee1** [comment]
@asvitkine I had forgotten your question 

_And on page 8, it's talking about Tank Armies and Army Corps which sound like they might be changing the number of dice rolled, etc. Wouldn't that not be supported with how Edit mode combat works due to number of dice differences?_

Yea the extra rolls show up in edit. So it works correctly.

Yea Paratroopers will work with no edit now. This can probably be closed. wc has opened a feature request at the site https://forums.triplea-game.org/topic/3797/paratroopers-and-u-boats

It'd still be nice to have at some point in the future. A lot easier to work with and other maps might use it differently.

And still keep edit as is. There are other situations it's still needed. The flexibility it provides is a great tool for non rule compliant games.

Anyway, Thanks again for your help.

**Comment by WCSumpton** [comment]
@beelee1 
You've created a condition to check every territory for paratroopers. Then created another condition to check every territory for the absence of allied units, except aaGuns. Created a trigger for every territory that checks those conditions to spawn a suicide unit to give a bonus for a single combat round. That's a wild hack! Way to go. LOL

Cheers...

**Comment by beelee1** [comment]
@WCSumpton heh heh not that hardcore :) What I did was add a "ParatrooperBoost" unit that gives the support. He's "isAir" with 99 movement, so can go anywhere. 
There are two place phases. So you need to buy as many as you need at no cost and then place in 1st place phase. That way there ready for combat that turn.

It's essentially Player Enforced. If you forget to buy one or not enough, need to edit one in at start of CM. They are "isConstruction" so they don't count as a placement.

You'll have to edit them in for defense after the battle if you win. I have some extra boxes too and have thought about just at top of map I could store some in. Then you could just click on them in NCM and "Perform Move or Other Actions" . Might be a little easier than adding by hand.
Still have to hand roll for any AA fire they encounter as the defender can choose his targets if normal Air Units attack with it. Otherwise I could make a special aaGun to shoot at just the Paratroopers. I suppose I could make one trhat only shoots once and you can edit those in for the defender as needed. 
Idk, it's not to bad to hand roll. Too bad Dice Roller wasn't documented in History :) We just use the A&A.org one for our PBF games.

The Wolfpack is the one I went big on :) But I already had that done for removing the Wolfpack if there were less than 3 U-boats. Just for Atlantic and Med. So I just copied it and changed that for it to add a Wolfpack when theres 3 or more :)

A little edit still but a Major improvement :)

**Comment by asvitkine** [comment]
So assuming we want to keep edit as-is, how to resolve the 10000 rounds error?

For battle simulation (that AI does behind the scenes), it's easy enough to just turn edit off for the simulations.
But then you get real AI battles - and those will just freeze triplea as it will actually try to go for that many rounds.

At first, I thought we could make AI retreat, but then that's not always an option.
I guess the question is who should be selecting casualties in edit mode in that case? If it's AI vs. AI, I guess we could just make casualties be selected as if it's not auto. But theoretically if it's an AI vs. human or human vs. AI battle, you do want casualties selected manually?

**Comment by asvitkine** [comment]
I think for now, I'll just make edit mode get turned off when transitioning to an AI player step. Before the change, that would have just hit that error anyway, so now at least it won't.

**Comment by asvitkine** [comment]
@beelee1 

I have another idea of how to make subs work as you want them. Same idea of spawning a special unit before attack phase whenever there are groups of three subs of the current player.

Make this unit suicide and make it actually have attacking rolls that the subs should have on first round. Then make its support attachment negatively affect the three subs by removing their dice rolls or changing what they hit on to 0.

I think that should work. Could you try it out?

**Comment by beelee1** [comment]
@asvitkine 

Yea Idk about player vs AI in edit. I would think that wouldn't work. Never tried it. I think edit would require a human to use it.

I'm actually quite pleased with how things are working now. Just a HUGE improvement thaks to your suggestion :) I like the support unit, because we can use hepps badass looking unit image, which helps tell at a glance that there is a wolfpack formed.

I will test out your latest idea though.

Thanks again. Really a Major improvement, with both it and paratroopers

**Comment by beelee1** [comment]
@asvitkine 

just reread. Yea, I see what you're saying now. The result would be the same, just a different way todo it. But I will test it

**Comment by beelee1** [comment]
@asvitkine 

Using your last suggestion I get an extra roll in combat. 3 U-boats and 1 Wolfpack get 4 rolls. Maybe I have the supportAttachment written wrong.

![Screenshot from 2024-05-08 14-33-37](https://github.com/triplea-game/triplea/assets/24382682/7742ccea-a26a-4066-bd09-b2778228b3d7)

![Screenshot from 2024-05-08 14-24-51](https://github.com/triplea-game/triplea/assets/24382682/4fd80aeb-a270-4d0a-92f0-9c6ccc96d4c1)

The negative bonus so the U-boats don't fire works correctly.
 
I tried it with 4 U-boats and got 5 dice. The Wolfpack is now A3 D3 and kills itself on both attack and defense.

Anyway, your original suggestion works great :) And revisiting this has made me realize I should try and have the Wolfpack spawn end of CombatTurn. Won't need to buy and place at all then :) 
I can also duplicate the triggers and have them respawn after place as well.

**Comment by asvitkine** [comment]
@beelee1 

Is the `Wolfpack` the unit that's spawned when there are three u-boats? Then the support attachment isn't set up how I expected.

I would expect it to be attached to `Wolfpack` and not `GermanUBoat` and for it to have `unitType` be `GermanUBoat`. So you want the Wolfpack affecting the UBoats, not the other way around.

Then you want `number` to be set to 3 to affect 3 uboats. And you set `	<option name='dice' value='strength'/>` and `<option name='bonus' value='-1'/>` to blank out their rolls.

**Comment by WCSumpton** [comment]
@beelee1, @asvitkine 

Right now, the global_40_expansion_uhd_boxes.xml is over 78k lines long and makes for a difficult read. Every 'work-around'/hack only make editing more difficult. It's no wonder @beelee1 would rather leave thing 'as they are'. Even though the hacks look good hear, each sea zone for subs, escorts, transports, etc. must be check 3 or 4 times per sea zone and there are 127 sea zones to check, and these checks are multiplied by each player. Without the use of variables and foreach loops, the number or xml lines just continues to increase.

Again, I suggest that adding "maxRounds" check to both supportAttachments and territoryEffectAttachments would be a better route for what @beelee1 wants to accomplish.

Cheers...

**Comment by beelee1** [comment]
@asvitkine 

> I would expect it to be attached to Wolfpack and not GermanUBoat and for it to have unitType be GermanUBoat. So you want the Wolfpack affecting the UBoats, not the other way around.

yea that's how I set it up the first time, which works. I was just trying your new idea because you asked me to. I don't really see the difference result wise.

The Wolfpack boosts the U-boats +1 for 1 rd in the first way. The second way the U-boats boost the Wolfpack for 1 rd with the Wolfpack taking away the U-boat 1st rd combat to 0.

The first way works good, so i'll just roll with that :)

@WCSumpton
yea it's pretty messy lol. I need to revisit the variables thing and try and clean it up. My past attempts have all failed :)

**Comment by asvitkine** [comment]
> @asvitkine
> 
> > I would expect it to be attached to Wolfpack and not GermanUBoat and for it to have unitType be GermanUBoat. So you want the Wolfpack affecting the UBoats, not the other way around.
> 
> yea that's how I set it up the first time, which works. 

I thought it didn't work on attack because a suicide unit dies before it can provide support? Or did I misunderstand.
I was trying to find a suggestion to workaround that.

> I was just trying your new idea because you asked me to. I don't really see the difference result wise.
I think there was a miscommunication since my new idea was still having the special unit to provide support. It was meant to workaround it not working on attack due to the suicide thing - but maybe I didn't understand and it actually works already?


> Again, I suggest that adding "maxRounds" check to both supportAttachments and territoryEffectAttachments would be a better route for what @beelee1 wants to accomplish.

Doesn't that still require the map script creating the support unit (which I think is the main part of the complexity)? I don't see it being different in terms of complexity of XML between the two solutions except adding that property requires coding changes whereas the other doesn't.

... And yeah if you want to do checks per-sea zone etc, that should be done with variables. If it's not doing that, that would be the first thing to clean up. (Just define a list of sea zones at the top of the file and do foreach on them.)

Also, just in terms of reducing file size, you can remove everything except the last name for all the `javaClass=` properties - i.e. `javaClass="UnitSupportAttachment"` instead of the very long thing. Same for every other attachment type.

**Comment by asvitkine** [comment]
> Also, just in terms of reducing file size, you can remove everything except the last name for all the `javaClass=` properties - i.e. `javaClass="UnitSupportAttachment"` instead of the very long thing. Same for every other attachment type.

I sent a PR for that here: https://github.com/triplea-maps/global_40_expansion_uhd_boxes/pull/4

**Comment by WCSumpton** [comment]
> Doesn't that still require the map script creating the support unit (which I think is the main part of the complexity)? I don't see it being different in terms of complexity of XML between the two solutions except adding that property requires coding engine changes (and a lot of complexity) whereas the idea behind my suggestion is existing features can already accomplish the task.

No, the GermanUBoat would apply the buff itself and there would be no need for a separate Wolfpack unit. If there was a "maxRounds" option in support attachments then:
~~~
    <attachment name="supportAttachmentWolfpackBuff" attachTo="GermanUBoat" javaClass="UnitSupportAttachment" type="unitType">
      <option name="unitType" value="GermanUBoat"/>
      <option name="faction" value="allied"/>
      <option name="side" value="offence"/>
      <option name="dice" value="strength"/>
      <option name="maxRounds" value="1"/>
      <option name="bonus" value="1"/>
      <option name="number" value="3"/>
      <option name="bonusType" value="Wolfpack_bonus" count="3"/>
      <option name="impArtTech" value="false"/>
      <option name="players" value="Germans"/>
    </attachment>
    <attachment name="supportAttachmentWolfpackDeBuff" attachTo="GermanUBoat" javaClass="UnitSupportAttachment" type="unitType">
      <option name="unitType" value="GermanUBoat"/>
      <option name="faction" value="allied"/>
      <option name="side" value="offence"/>
      <option name="dice" value="strength"/>
      <option name="maxRounds" value="1"/>
      <option name="bonus" value="-1"/>
      <option name="number" value="2"/>
      <option name="bonusType" value="Wolfpack_Debuff" count="2"/>
      <option name="impArtTech" value="false"/>
      <option name="players" value="Germans"/>
    </attachment>
~~~
Is all the code needed to handle the attack support. It's like "maxRoundsAA", which govern how may rounds a unit may conduct an AA attack.

Cheers...

**Comment by beelee1** [comment]
@asvitkine 

yea muts've been a Led Zeppelin  breakdown. :) 
Your first solution works great. 

I will look at the suggestions to clean up the code.

@WCSumpton 

yea still wanna have the Wolfpack image show up thow, which is easy enough todo with conditions.

Anyway, it's working real good right now. I just changed it so the Wolfpacks place or remove before combat. Also after place. For placement, so if you buty 3 and place, then it'll spawn as well. 

Only time we'd need to edit is if the Wolfpack disolved in defense.
Like USA attacks, kills a U-boat and then UK attacks right after. If there were still 3 or more Wolfpacks left after the USA attack, you'd have to edit the Wolfpack back in since he'd be dead from the combat.

Anyway, I don't wanna bug on your guy's time. It's working real good right now.

Thanks for all your help :)

[
![Rock On !!!](https://github.com/triplea-game/triplea/assets/24382682/bb971eeb-434c-48d5-9ed3-b4eb2fe444a3)
](url)

**Comment by beelee1** [comment]
@WCSumpton 

Actually your suggestion is the best, because if the U-boats were attacked during another players turn and still had at leat 3 present, the behavior would be correct.

The image would be gone, but bfd it'll respawn :)

**Comment by WCSumpton** [comment]
@beelee1 

Let's take a second to look at what @asvitkine is suggesting. First you would need to create variable list or all the sea zones:
~~~
<variableList>
   <!-- A list of all Atlantic/Mediterranean sea zones to check for Walfpack -->
   <!-- There are 52 sea zones to check -->
   <variable name="WolfpackSeaZones">
      <!-- All sea zones end in " Sea Zone" so only the number is set -->
      <element name="127"/>
      <element name="126"/>
      ...
      ...
      ...
      <element name="68"/>
      <element name="67"/>
   </variable>
</variableList>
~~~
Now for some conditions:
~~~
<!-- Check for the presence of 3 German U-Boats -->
<attachment foreach="$WolfpackSeaZones$" name="conditionAttachment_3_GermanUBoat_@WolfpackSeaZones@_SeaZones" attachTo="Germans" javaClass="RulesAttachment" type="player">
   <option name="directPresenceTerritories" value="@WolfpackSeaZones@ Sea Zone" count="1"/>
   <option name="unitPresence" value="GermanUBoat" count="3"/>
</attachment>
<!-- Also check for the presence of any other friendly unit -->
<!-- This should check for all sea and air unit that can aid -->
<attachment foreach="$WolfpackSeaZones$" name="conditionAttachment_No_Allied_Help_@WolfpackSeaZones@_SeaZones" attachTo="Germans" javaClass="RulesAttachment" type="player">
   <option name="directPresenceTerritories" value="@WolfpackSeaZones@ Sea Zone" count="1"/>
   <option name="unitPresence" value="Battleship:Destroyer:every thing but Transport:and all air" count="1"/>
   <option name="invert" value="true"/>
</attachment>
~~~
Next are the triggers:
~~~
<!-- Place Wolfpack where there are 3 German U-Boats with no allied help -->
<attachment foreach="$WolfpackSeaZones$" name="triggerAttachment_Wolfpack_at@WolfpackSeaZones@_SeaZones" attachTo="Germans" javaClass="TriggerAttachment" type="player">
      <option name="conditions" value="conditionAttachment_3_GermanUBoat_@WolfpackSeaZones@_SeaZones:conditionAttachment_No_Allied_Help_@WolfpackSeaZones@_SeaZones"/>
      <option name="placement" value="@WolfpackSeaZones@ Sea Zone:Wolfpack" count="1"/> 
      <option name="when" value="after:germansCombatMove"/>
      <!-- These are added to ensure Wolfpack is placed for defensive purposes -->
      <option name="when" value="after:russiansCombatMove"/>
      <option name="when" value="after:americansCombatMove"/>
      <option name="when" value="after:chineseCombatMove"/>
      <option name="when" value="after:britishCombatMove"/>
      <option name="when" value="after:anzacCombatMove"/>
      <option name="when" value="after:frenchCombatMove"/>
    </attachment>
    <!-- Remove Wolfpack prior to NonCombat move (There was no check for enemy units) -->
    <attachment name="triggerAttachment_Remove_All_Wolfpack" attachTo="Germans" javaClass="TriggerAttachment" type="player">
      <option name="conditions" value="conditionAttachment_SwitchIsTrue"/>
      <!-- There are 52 sea zones, so 55 gives a little cushion -->
      <option name="removeUnits" value="all:Wolfpack" count="55"/> 
      <option name="when" value="before:germansNonCombatMove"/>
      <option name="when" value="before:russiansNonCombatMove"/>
      <option name="when" value="before:americansNonCombatMove"/>
      <option name="when" value="before:chineseNonCombatMove"/>
      <option name="when" value="before:britishNonCombatMove"/>
      <option name="when" value="beforeanzacNonCombatMove"/>
      <option name="when" value="before:frenchNonCombatMove"/>
    </attachment>
~~~
If this seems easier, then there you have it!

Cheers...

**Comment by beelee1** [comment]
@WCSumpton 

Oh way sweet you taking time on this. :)

I'll look closer tommorow as it's getting late in the day for me :)

But yea, i need to get that variable shit to work lol

**Comment by asvitkine** [comment]
@WCSumpton 
 
> No, the GermanUBoat would apply the buff itself and there would be no need for a separate Wolfpack unit. If there was a "maxRounds" option in support attachments then:

I'm not really following how this works, do you mind explaining? How does it ensure that the bonus only applies when there are three units present?

I see you're using `count` on `bonusType`, but I don't see how that helps here, from the docs:

```
Can use count (defaults to 1) to stack the same bonusType on a single unit like <option name="bonusType" value="ArtStackBonus" count="3"/> where a unit could receive up to 3 units supporting it of that bonusType. Any support of the same bonusType MUST have the same count.
```

**Comment by WCSumpton** [comment]
@asvitkine 

A unit can give a support to the supported unit once
A unit can give multiple supports to the same unit as long as each support is named differently 
GermanUBoats can apply a +1 bonus to 3 (number) GermanUBoats and that bonus can stack 3 times (count for bonusType Wolfpack_bonus). It can also apply a -1 bonus to 2 GermanUBoats that can only stack twice.
1 GermanUBoat attacks at 2 +1 Wolfpack_bonus -1 Wolfpack_Debuff = 2 Attack
2 GermanUBoats attacks at 2 +2 Wolfpack_bonus -2 Wolfpack_Debuff = 2 Attack each
3 GermanUBoats attacks at 2 +3 Wolfpack_bonus -2 Wolfpack_Debuff = 3 Attack each

Hope this explains what I am talking about.

Cheers...

**Comment by Cernelius** [comment]
Beside that the "number" of the second attachment should be 3 too, have you tested it (asking because I haven't)?

To make a simpler example, are you sure that, when there is a unit which can give support to 3 other units and each of these units can receive up to 3 supports of that type, each of the units are supported once instead of only one of the units being supported thrice?

I agree that should be the behaviour (and I suggest to change it to that if it is not), but I remember that I argued about it with @ron-murhammer, and I remember that he instead decided to go for stacking supports on the same unit as much as possible before supporting more as a behaviour.

Again, I've not tested this specifically, and it was years ago, so I may be remembering wrong. Testing is euncouraged.

**Comment by WCSumpton** [comment]
@Cernelius 

I have tested this. Also, every unit added will receive that 'last bonus'. So, even though the bonus stopes at 3, adding a fourth unit and all 4 unit will attack at 3. As to the debuff, stack count is 2, and it is the same unit, so the number does not matter as long as it is equal or greater than the stack count.

Cheers...

**Comment by Cernelius** [comment]
The number I was referring to is the "number", not the count of the "bonusType".

**Comment by WCSumpton** [comment]
@Cernelius 

That is what I said, the "number" does not matter as long as it is equal to or greater than the stack count. 2, 3, 4 they work because it is the unit debuffing itself, so the addition of another unit only adds to the stack count not the "number" of unit affected. Funny things will happen if the "number" is less than the stack count. LOL

Cheers...

**Comment by Cernelius** [comment]
Yes, I see it now, and anyway you tested it so most likely all good.

Still, if what I remember from years ago is correct, that would mean that, when a single unit can support multiple units each of which can be supported multiple times by that same type of support, the behaviour is never to support any same unit more than once even if that means wasting the exceeding supporting modifiers, whereas, when each of multiple units can support a single unit, the behaviour is to support a same unit as much as possible before supporting more units. Am I right (Still have not tested it yet.)? If so, that seems somewhat inconsistent to me, but I need to get around doing some testing about it and maybe opening a discussion or an issue about it.

---

Anyway, since you tested it that it works that way, I definitely agree that having a "maxRounds" is all that it is needed, and doing it that way is way better and cleaner. On top of that, a "maxRounds" would be a very valuable feature on its own.

**Comment by Cernelius** [comment]
It is also an interesting albeit rough example of sort of "economies of scale" (not the right phrase), where having more submarines in the same zone makes each of them more effective.

It makes me think to some cards of "Magic" where the offence/defence (or whatever it is called) of each of the cards is equal to the number of these cards in play.

In TripleA terms, it would be like that, if you have 1 unit in the battle, the unit is 1/1, if you have 2 units in the battle, each of the units is 2/2, if you have 3 units in the battle, each of the units is 3/3, and so on.

**Comment by beelee1** [comment]
@WCSumpton 

So I get this error when I start the game. I think that's the same one I've gotten in the past when trying to add variables. 

![Screenshot from 2024-05-16 17-51-51](https://github.com/triplea-game/triplea/assets/24382682/7e774570-e993-47c8-a4f2-c1deb0131608)

I removed the existing Wolfpack triggers and added this to the xml. The Wolfpacks are active whether Allied units are present or not, so I removed that condition from the trigger.
That will work good for the Paratroopers though :)

Hmm ... paste didn't seem to work. Guess I'll add some shots.

![Screenshot from 2024-05-16 18-02-25](https://github.com/triplea-game/triplea/assets/24382682/bb9505c8-db2f-48ad-8e8e-ad394d08af7f)

![Screenshot from 2024-05-16 18-02-48](https://github.com/triplea-game/triplea/assets/24382682/7cbf676f-1214-4659-899a-d2258a3e8ea6)

![Screenshot from 2024-05-16 18-03-11](https://github.com/triplea-game/triplea/assets/24382682/9c94e2ad-e823-4dcc-8258-19c5a9005385)

![Screenshot from 2024-05-16 18-04-07](https://github.com/triplea-game/triplea/assets/24382682/f696a06a-e60f-4e88-a1a6-681fad7deedf)

**Comment by WCSumpton** [comment]
@beelee1 

If I am reading this correctly, you are creating your variable list in the wrong location inside the xml. The variable list needs to begin at line 8 in your xml. Also, the variable list at line 74340 through 74373 needs to be removed. So, starting at line 8 your variable list should end at line 64 with the <map> starting at line 65

**Comment by beelee1** [comment]
@WCSumpton  No joy yet. Variable list starts at L-8 and deleted the others. Conditions at L-15053 and triggers at L-44579.

I removed that 2nd condition from the trigger. I don't think it should impact anything but I will test it with it included next.

Here is xml

[EXP 2.13 5-17-24.zip](https://github.com/triplea-game/triplea/files/15358546/EXP.2.13.5-17-24.zip)

**Comment by beelee1** [comment]
@WCSumpton hmm I added this condition to the "remove" trigger and get a error again.

<attachment foreach="$WolfpackSeaZones$" name="conditionAttachment_3_GermanUBoatRemove_@WolfpackSeaZones@_SeaZones" attachTo="Germans" javaClass="RulesAttachment" type="player">
   <option name="directPresenceTerritories" value="@WolfpackSeaZones@ Sea Zone" count="1"/>
   <option name="unitPresence" value="GermanUBoat" count="3"/>
   <option name="invert" value="true"/>
</attachment>

The remove trigger was removing them all. It's only condition was "switch is true" so added the above which is just the invert of the placement one.


    <attachment name="triggerAttachment_Remove_All_Wolfpack" attachTo="Germans" javaClass="TriggerAttachment" type="player">
      <option name="conditions" value="conditionAttachment_SwitchIsTrue:conditionAttachment_3_GermanUBoatRemove_@WolfpackSeaZones@_SeaZones"/>
      
      <option name="removeUnits" value="all:Wolfpack" count="59"/> 
      <option name="when" value="before:germansNonCombatMove"/>
      <option name="when" value="before:russiansNonCombatMove"/>
      <option name="when" value="before:americansNonCombatMove"/>
      <option name="when" value="before:chineseNonCombatMove"/>
      <option name="when" value="before:britishNonCombatMove"/>
      <option name="when" value="before:anzacNonCombatMove"/>
      <option name="when" value="before:frenchNonCombatMove"/>
    </attachment>

**Comment by WCSumpton** [comment]
@beelee1 
<map> at line 8 needs to be moved to line 67.
Line 15062 you should add all sea unit like carrier, cruiser, Escort, Escort_Convoy, German_Heavy_BB also destroyer and battleship should not be capitalized.

The second condition is the one that checks for the existence of other German sea units. It was my understanding that for the Wolfpack bonus there should be 3 German_UBoats and no other surface warship presence. If that is not the case then the condition at line 15060 can be removed, as that is what this condition checks.

Cheers...

**Comment by beelee1** [comment]
@WCSumpton uggh Idky git won't let me paste wtf. Here's the condition. Or maybe not jfc

![Screenshot from 2024-05-17 18-12-06](https://github.com/triplea-game/triplea/assets/24382682/bbb865a6-9c3f-4026-8a9d-69a469ea7df2)


<attachment foreach="$WolfpackSeaZones$" name="conditionAttachment_3_GermanUBoatRemove_@WolfpackSeaZones@_SeaZones" attachTo="Germans" javaClass="RulesAttachment" type="player">
   <option name="directPresenceTerritories" value="@WolfpackSeaZones@ Sea Zone" count="1"/>
   <option name="unitPresence" value="GermanUBoat" count="3"/>
   <option name="invert" value="true"/>
</attachment>

**Comment by beelee1** [comment]
@WCSumpton yea I caught that for map and got it to fire. Yea the Paratroopers are the ones that get negated by others. The Wolfpacks get there bonus regardless of other units being present.

Trying to get my remove trigger to work now.

**Comment by WCSumpton** [comment]
@beelee1 

Looks like you have an extra condition in your remove trigger. There should be only the always true condition.

Cheers...

**Comment by beelee1** [comment]
that removes them when it shouldn't. As long as there are 3 U-boats they shouldn't be removed. Hmm ... maybe I need to play it out and see if they respawn, but I'd prefer they not remove as long as there are 3 or more present

**Comment by WCSumpton** [comment]
@beelee1 

The question should be, when should the WolfPack unit be updated. I went for the quick fix, remove the before non-combat movement, and then respawn them prior to a combat round. What you want is to have them visually available all the time.
So, something like this:
~~~
<!--Next are the triggers: Place Wolfpack where there are 3 German U-Boats -->
<attachment foreach="$WolfpackSeaZones$" name="triggerAttachment_Wolfpack_at@WolfpackSeaZones@_SeaZones" attachTo="Germans" javaClass="TriggerAttachment" type="player">
  <option name="conditions" value="conditionAttachment_3_GermanUBoat_@WolfpackSeaZones@_SeaZones"/>
  <option name="placement" value="@WolfpackSeaZones@ Sea Zone:Wolfpack" count="1"/> 
  <option name="when" value="before:germansBattle"/>
  <option name="when" value="before:germansNonCombatMove"/>
  <option name="when" value="before:germansPurchase"/>
  <option name="when" value="before:germansEndTurn"/>
  <!-- These are added to ensure Wolfpack is replaced after combat -->
  <option name="when" value="before:russiansNonCombatMove"/>
  <option name="when" value="before:americansNonCombatMove"/>
  <option name="when" value="before:chineseNonCombatMove"/>
  <option name="when" value="before:britishNoAirCheckPlace"/>
  <option name="when" value="before:anzacNonCombatMove"/>
  <option name="when" value="before:frenchNonCombatMovee"/>
</attachment>
<!-- Remove Wolfpack a combat round, movement or placement phase -->
<attachment name="triggerAttachment_Remove_All_Wolfpack" attachTo="Germans" javaClass="TriggerAttachment" type="player">
  <option name="conditions" value="conditionAttachment_SwitchIsTrue"/>
  <!-- There are 52 sea zones, so 55 gives a little cushion -->
  <option name="removeUnits" value="all:Wolfpack" count="59"/> 
  <option name="when" value="after:germansCombatMove"/>
  <option name="when" value="after:germansBattle"/>
  <option name="when" value="after:germansNonCombatMove"/>
  <!-- will cause problems because of multiple phases with the same name -->
  <option name="when" value="after:germansPlace"/>
  <option name="when" value="after:russiansBattle"/>
  <option name="when" value="after:americansBattle"/>
  <option name="when" value="after:chineseBattle"/>
  <option name="when" value="after:britishBattle"/>
  <option name="when" value="after:anzacBattle"/>
  <option name="when" value="after:frenchBattle"/>
</attachment>
~~~

Cheers...

**Comment by beelee1** [comment]
@WCSumpton That works, although Idky it needs to remove and then place again. I thought my remove condition would only remove when less than 3 are present, but that obviously didn't work.

Still don't quite understand how it all works, but the important thing is it does :)

Not sure what you mean here:

_will cause problems because of multiple phases with the same name_



Because I have two place phases ? Is it going to affect gameplay ?

Thanks again for the help :)

**Comment by WCSumpton** [comment]
> I agree that should be the behaviour (and I suggest to change it to that if it is not), but I remember that I argued about it with @ron-murhammer, and I remember that he instead decided to go for stacking supports on the same unit as much as possible before supporting more as a behaviour.

@Cernelius 

You are right the count has been changed. According to PoS2:
`bonusType	values: the name of the bonus, any string. Any bonus with the same name will not stack unless a count of how many can stack is specified, any bonus with a different name will stack. 
			Can use count (defaults to 1) to stack the same bonusType on a single unit like <option name="bonusType" value="ArtStackBonus" count="3"/> where a unit could receive up to 3 units supporting it of that bonusType. Any support of the same bonusType MUST have the same count. 
			A count of -1 indicates the support can stack infinitely and a count of 0 means it will not apply at all.
			Please us the string "ArtyOld" if do not want your bonus to stack with normal artillery (ie: do not want the bonus to stack with any unit with the normal artillery attachment: <option name="artillery" value="true"/>)
`

And the line **Any support of the same bonusType MUST have the same count.** is no longer true.

I'm not sure what count does now, but it no longer reflects what is documented.

Thanks for the heads up. 

So, the question remains, what does count do?

Cheers...

**Comment by WCSumpton** [comment]
> Because I have two place phases ? Is it going to affect gameplay ?

@beelee1 

Don't know much about gameplay, @TheDog-GH could help better with that. It will mess up triggers that have the "when" before/after xxxPlace. Some may fire twice, some may fire once and other may not work at all. There are 136 triggers that use germansPlace alone.

Cheers...

**Comment by TheDog-GH** [comment]
@beelee1 
This map is complex, well done!
I agree with WCSumpton, for his stated reasons, you should change the label of the 2nd Place to have a say a 2 in them so
```
line 1821   <step name="germansPlace" delegate="place" player="Germans"/>
change to  <step name="germans2Place" delegate="place" player="Germans"/>
                              ^
```
I would do this for all nations that have 2x Place steps.

This way the engine will not confuse the two Place and all your current triggers will fire on the 1st Place, so they should work as intended.  Unless they are meant to fire on **2**Place.

**Comment by beelee1** [comment]
@TheDog-GH oh that'll work great ! I didn't know you could give them different names. :)

**Comment by WCSumpton** [comment]
@beelee1 

Check out Invasion USA, the Americans have 5 tech/techActivation steps. So, you can have as many steps as you want. But to insure triggers fire at the right time, steps should have unique names.

Cheers...

**Comment by beelee1** [comment]
@TheDog-GH 

So i have certain units that are purchased that can be placed in the first place phase and others only in the 2nd. 

Example
Germans may purchase Waffen Infantry and regular Infantry. I would like only the Waffen Infantry to be able to place the 1st place phase and then Wehrmacht Infantry the second. Both are purchased at same time.

Right now it is just player enforced and seems to work well enough once players are familiar with the rules. I guess two purchase phases maybe ? That only allows certain units to be purchased ?

That would be overly complex and I don't think triplea could do it anyway.

@WCSumpton Thank You I will check that out.

Hmm ... rereading your comment Dog, so I could add a 2nd purchase phase and then have those units become available for purchase via trigger.

So it would go, limited purchase and place and then a 2nd purchase, which triggers more units to be purchased and then they could be placed in the 2nd place phase.

That might work, but seems quite messy.  Probably better to player enforce but worth experimenting with :)

**Comment by beelee1** [comment]
@WCSumpton heh heh 

yea I think it might not be worth the effort, but if triplea can do it, that's the question :)

**Comment by WCSumpton** [comment]
@beelee1 

Yes, it can, there is the Bid stuff. Right now, I'm playing around with a Bid purchase, Pre-purchase idea.

Cheers...

**Comment by beelee1** [comment]
@WCSumpton let us know how it goes :)

I've got another one but I'm gonna beat my head on the wall a bit before i ask :)

**Comment by TheDog-GH** [comment]
@beelee1 
What about this variant,
Give 1 or 2 unit per turn of say Panzer_Grndrs using eg.
<option name="purchase" value="Panzer_Grndrs" count="2"/>

Also for upgrading units is messy, use the same idea for WaffenPnzrGrndr, just give the unit

So in english, just drip feed the player with 1 or 2 units per turn with triggers to be placed in the 2Place phase
so no 2Purchase phase, 
no upgrading/replacing units with consumesUnits
no maxBuiltPerPlayer (as they are drip fed)
Also the AI can cope with the above

You can give the units for free or if you want the player to pay, then the 2Place, needs to be before Place (so swap them) so that the auto purchase (2Place) has enough PU to pay for the scripted units.

**Comment by beelee1** [comment]
@TheDog-GH 

yea thanks Dog. This is a mod the Danish A&A Group came up with after 25 years of testing, adding and subtracting stuff, so I need to make it as rule compliant as possible to what they have.

I'm not really part of the design group for it :) Just looked cool and wanted to see if triplea could do it. While not perfect, it's actually pretty close.

Player enforced play is not as big a issue as edit imo. Just makes it play more like a tabletop game :)

Anyway, you and WC have givin me enough to stay busy for a while, as it continues to improve :)
